formal-spec/Leios/Linear.lagda.md

```diff
@@ -134,16 +140,22 @@ mempool.
           ∙ validityCheckTime eb ≤ 3 * Lhdr + Lvote
           ∙ EndorserBlockOSig.txs eb ≢ []
           ∙ needsUpkeep VT-Role
-          ∙ canProduceV (slotNumber eb) sk-VT (stake s)
+          ∙ canProduceV slot sk-VT (stake s)
           ───────────────────────────────────────────────────────
           s ↝ ( rememberVote (addUpkeep s VT-Role) eb
-              , Send (vtHeader [ vote sk-VT (hash eb) ]) nothing)
+              , just (Send (vtHeader [ vote sk-VT (hash eb) ]) nothing))
+
+  No-VT-Role : let open LeiosState s in
```

Comment:
The reason I removed the negative votes was that we need to be careful to not get stuck. This can either be done by `_↝_` being total (which means negative cases) or by still making a step if `_↝_` cannot.

The issue is that to make `_↝_` total is non-trivial here, since there are lots of reasons why `VT-Role` cannot make a step. E.g. the EB could be equivocated or the header could have arrived too late. So if we add the negative cases, they are going to be quite complicated and we should include a proof that `_↝_` is indeed total.

This is why I removed the negative cases, since it's just so much extra work. Do you need them for something?

Reply:
Ok, I see, we would need all the different negative rules not the get stuck, in particular for `VT-Role`. For trace verification, this would actually be convenient, but I agree, it somehow bloats the spec